### PR TITLE
Feature/checksum pipeline

### PR DESCRIPTION
**Pull request by vinay-ebi**
## Description
New pipeline to generate secure hash  using sha512 and md5 for genomic sequences (toplevel, cda, cdna and peptide) and add the generated hash value into respective attribute tables

## Benefits
Generate secure hash for individual species and sequence type

More details added into the ticket : https://www.ebi.ac.uk/panda/jira/browse/ENSPROD-8593

### CODE REVIEW

**Review by andrewyatz — CHANGES_REQUESTED** [review]
I think there's a few bits of the code which can be

- Simplified by not passing around subroutines unnecessarily 
- Made more memory efficient by trying to handle large sequences more appropriately i.e. pass by reference
- Better use of defaults using eHive paradigms vs. inline defaults
- Correctly referring to the algorithms being used i.e. `md5` and `sha512t24u`

File: modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/CheckProductionDumps_conf.pm [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+
+  return {
+    %{$self->SUPER::pipeline_wide_parameters},
+    'pipeline_name' => $self->o('pipeline_name'),
+    'base_path'     => $self->o('ftp_dir'),
+    'release'       => $self->o('release'),
+  };
+}
+
+sub resource_classes {
```
Comment (marcoooo) [inline modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/CheckProductionDumps_conf.pm:157]: overriding resource class would prevent us from using this pipeline, say with SLURM. 
Can't we just integrate those at a higher level in the hierarchy?

File: modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/CheckProductionDumps_conf.pm [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package Bio::EnsEMBL::Production::Pipeline::PipeConfig::DumpCore_conf;
```
Comment (marcoooo) [inline modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/CheckProductionDumps_conf.pm:20]: Package doesn't correspond to file name, btw, is this file related to the Checksum ticket at all?

File: modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/Ga4gh_conf.pm [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package Bio::EnsEMBL::Production::Pipeline::PipeConfig::Ga4gh_conf;
```
Comment (marcoooo) [inline modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/Ga4gh_conf.pm:20]: I would personally not call it ga4gh, this is a ChecksumGenerator, which might be used in other contexts

File: modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/Ga4gh_conf.pm [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    'pipeline_name' => $self->o('pipeline_name'),
+    'release'       => $self->o('release'),
+    'sequence_types'         => $self->o('sequence_type'),
+    'hash_types'             => $self->o('hash_type'),
+
+
+  };
+}
+
+sub resource_classes {
```
Comment (marcoooo) [inline modules/Bio/EnsEMBL/Production/Pipeline/PipeConfig/Ga4gh_conf.pm:85]: Same comment about the resource class, it doesn't make any sense to add things here, better to put it at top level in hierarchy, so we can use this pipeline on SLURM

**Review by marcoooo — CHANGES_REQUESTED** [review]
Can we renamed the Ga4gh package to something more generic please?

**Review by marcoooo — APPROVED** [review]

**Review by marcoooo — APPROVED** [review]

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+sub run {
+
+  my ($self) = @_;
+  my $hash_types = $self->param('hash_types');
+  my $type = $self->param('group');
+  my $species = $self->param('species');
+  my $seq_type = $self->param('seq_type');
+  my ($genome_ad, $genome_feature,  $sequence, $attrib_table, $genome_feature_id, $genome_slice) ;
+  my $dba  = $self->get_DBAdaptor($type);
+  my $registry = 'Bio::EnsEMBL::Registry';
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:101]: Why do we bring in the registry here?

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+sub run {
+
+  my ($self) = @_;
+  my $hash_types = $self->param('hash_types');
+  my $type = $self->param('group');
+  my $species = $self->param('species');
+  my $seq_type = $self->param('seq_type');
+  my ($genome_ad, $genome_feature,  $sequence, $attrib_table, $genome_feature_id, $genome_slice) ;
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:99]: `$genome_slice` seems an odd name when a lot of things below are not slices ...

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    my $at_code = $hash_method."_".$seq_type;
+    my $sequence_hash = $self->generate_sequence_hash($genome_slice, $hash_method);
+    $self->update_attrib_table($attribute_adaptor, $at_code, $sequence_hash, $genome_feature_id , $attrib_table);
+  }  
+}
+
+sub generate_sequence_hash {
+  my ($self, $slice, $hash_method) = @_;
+  my $sequence_hash;
+  if( $hash_method =~ /sha\-?(\d+)/){
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:143]: This seems unnecessary to me. If we have another algorithm e.g. `sha512t30u`, then wouldn't we specifically call that out rather than parsing out something which will a the moment always be 24?

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+  return $sequence_hash; 
+}
+
+sub sha512t24u {
+  my ($self, $slice, $algo, $digest_size) = @_;
+
+  if(($digest_size % 3) != 0) {
+    die "Digest size must be a multiple of 3 to avoid padded digests";
+  }
+  my $sha = Digest::SHA->new(int($algo)); #even though we use sha512 algorithm we truncate the generated hash based on given digest size
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:159]: Hard code IMO

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+sub md5{
+  my ($self, $slice) = @_;
+  my $md5 = Digest::MD5->new;
+  $md5 = $self->sequence_stream_digest($slice, $md5, $self->param('chunk_size'));
+  my $digest = $md5->digest;
+  my $base64 = encode_base64url($digest);
+  return $base64 
+}
+
+sub sequence_stream_digest {
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:176]: Just double checking that this is working correctly i.e. we can push say the MD5 checksums into ENA's refget and find our sequences?

**Review by andrewyatz — CHANGES_REQUESTED** [review]
Few comments in there which I think is making the code more complicated than it needs to be IMO

**Comment by andrewyatz** [comment]
Also on that last point we are sure this is working correct ...

**Comment by vinay-ebi** [comment]
> Also on that last point we are sure this is working correct ...
Tested complete pipeline for species actinobacillus_pleuropneumoniae_serovar_5b_str_l20_gca_000015885
http://guihive-hl.ebi.ac.uk:8080/versions/96/?driver=mysql&username=ensadmin&host=mysql-ens-hive-prod-1&port=4575&dbname=vinay_ga4gh_test_110&passwd=xxxxx

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+  my ($self, $slice) = @_;
+  my $md5 = Digest::MD5->new;
+  $md5 = $self->sequence_stream_digest($slice, $md5, $self->param('chunk_size'));
+  my $digest = $md5->hexdigest;
+  return $digest
+}
+
+sub sequence_stream_digest {
+  my ($self, $slice, $hash_obj, $chunk_size) = @_;
+  $chunk_size = $self->param('$chunk_size') unless $chunk_size;
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:177]:
```suggestion
  my $chunk_size = $self->param('chunk_size') unless $chunk_size;
```

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+sub md5{
+  my ($self, $slice) = @_;
+  my $md5 = Digest::MD5->new;
+  $md5 = $self->sequence_stream_digest($slice, $md5, $self->param('chunk_size'));
+  my $digest = $md5->hexdigest;
+  return $digest
+}
+
+sub sequence_stream_digest {
+  my ($self, $slice, $hash_obj, $chunk_size) = @_;
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:176]:
```suggestion
  my ($self, $slice, $hash_obj) = @_;
```

**Review by andrewyatz — CHANGES_REQUESTED** [review]

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+  }elsif($hash_method eq "md5"){
+     $sequence_hash = $self->md5($slice);	  
+  }else{
+     die "Unknown hash method $hash_method, valid hash methods ex: 'sha512t(24|30..)u?', 'md5'";	  
+  }	  
+  return $sequence_hash; 
+}
+
+sub sha512t24u {
+  my ($self, $slice, $algo, $digest_size) = @_;
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:154]:
```suggestion
  my ($self, $slice) = @_;
```

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+  if(($digest_size % 3) != 0) {
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:156]:
```suggestion
my $digest_size = 24;
```

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    my $at_code = $hash_method."_".$seq_type;
+    my $sequence_hash = $self->generate_sequence_hash($sequence_obj, $hash_method);
+    $self->update_attrib_table($attribute_adaptor, $at_code, $sequence_hash, $genome_feature_id , $attrib_table);
+  }  
+}
+
+sub generate_sequence_hash {
+  my ($self, $slice, $hash_method) = @_;
+  my $sequence_hash;
+  if( $hash_method =~ /sha\-?(\d+)t(\d+)u?/){
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:141]:
```suggestion
  if( $hash_method eq 'sha512t24u'){
```
I believe it is easier to ignore all of this parsing and just select based on the hash method name. We have two valid hash methods, `md5` and `sha512t24u`. Nothing else at the time of writing is valid

File: modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    my $digest_size = $self->param('digest_size') || 24;
+    my $algo = int($1);    
+    $sequence_hash = $self->sha512t24u($slice, $algo, $digest_size);	  
+  }elsif($hash_method eq "md5"){
+     $sequence_hash = $self->md5($slice);	  
+  }else{
+     die "Unknown hash method $hash_method, valid hash methods ex: 'sha512t(24|30..)u?', 'md5'";
```
Comment (andrewyatz) [inline modules/Bio/EnsEMBL/Production/Pipeline/Ga4ghChecksum/CheckSumGenerator.pm:148]: Why not hard-code valid hashes in the params method? It would potentially simplify the reporting here

**Review by andrewyatz — CHANGES_REQUESTED** [review]
Some more changes sorry. I really believe you should reduce the complexity of what's being looked at here. The code is more complicated than it needs to be in my opinion

**Review by andrewyatz — APPROVED** [review]
Looks good and cleaner. I can also see the use of registry later on. Thanks @vinay-ebi

**Comment by vinay-ebi** [comment]
> Looks good and cleaner. I can also see the use of registry later on. Thanks @vinay-ebi

Thanks @andrewyatz  , for the valuable suggestion and refining the code
